Firestore/core/src/firebase/firestore/remote/grpc_stream.h

```diff
@@ -0,0 +1,238 @@
+
+  // Does not produce a notification. Once this method is called, the stream can
+  // no longer be used.
+  //
+  // Can be called on a stream before it opens. It is invalid to finish a stream
+  // more than once.
+  void Finish();
+
+  /**
+   * Writes the given message and finishes the stream as soon as the write
+   * succeeds. Any non-started writes will be discarded. Neither write nor
+   * finish will notify the observer.
+   *
+   * If the stream hasn't opened yet, `WriteAndFinish` is equivalent to
+   * `Finish` -- the write will be ignored.
+   */
+  void WriteAndFinish(grpc::ByteBuffer&& message);
+
+ private:
+  friend class internal::GrpcStreamDelegate;
+
+  GrpcStream(std::unique_ptr<grpc::ClientContext> context,
+             std::unique_ptr<grpc::GenericClientAsyncReaderWriter> call,
+             GrpcStreamObserver* observer,
+             GrpcCompletionQueue* grpc_queue);
+
+  void Read();
+  void BufferedWrite(grpc::ByteBuffer&& message);
+
+  // Called by `GrpcStreamDelegate`.
+  void OnStart();
+  void OnRead(const grpc::ByteBuffer& message);
+  void OnWrite();
+  void OnOperationFailed();
+  void OnFinishedByServer(const grpc::Status& status);
+  void OnFinishedByClient();
+  void RemoveOperation(const internal::StreamOperation* to_remove);
+
+  // Whether this stream belongs to the same generation as the observer.
+  bool SameGeneration() const;
+
+  // Creates an operation, stores a `unique_ptr` to it, and returns a plain
+  // pointer.
+  template <typename Op, typename... Args>
+  Op* MakeOperation(Args... args);
+
+  // Creates and immediately executes an operation.
+  template <typename Op, typename... Args>
+  void Execute(Args... args) {
+    MakeOperation<Op>(args...)->Execute();
+  }
+
+  // The gRPC objects that have to be valid until the last gRPC operation
+  // associated with this call finishes. Note that `grpc::ClientContext` is
+  // _not_ reference-counted.
+  //
+  // Important: `call_` has to be destroyed before `context_`, so declaration
+  // order matters here. Despite the unique pointer, `call_` is actually
+  // a non-owning handle, and the memory it refers to will be released once
+  // `context_` (which is owning) is released.
+  std::unique_ptr<grpc::ClientContext> context_;
+  std::unique_ptr<grpc::GenericClientAsyncReaderWriter> call_;
+
+  GrpcCompletionQueue* grpc_queue_ = nullptr;
+
+  GrpcStreamObserver* observer_ = nullptr;
+  int generation_ = -1;
+  // Buffered writer is created once the stream opens.
+  absl::optional<BufferedWriter> buffered_writer_;
+
+  std::vector<std::unique_ptr<internal::StreamOperation>> operations_;
+
+  // The order of stream states is linear: a stream can never transition to an
+  // "earlier" state, only to a "later" one (e.g., stream can go from `Started`
+  // to `Open`, but not vice versa). Intermediate states can be skipped (e.g.,
+  // a stream can go from `Started` directly to `Finishing`).
+  enum class State {
+    NotStarted,
+    Started,
```

Comment:
What's the difference between Started and Open? Maybe you mean Starting here?

Reply:
Changed to `Starting`.